core: Implement gracefull shutdown for a brick process

mohit84 opened the pull request:
[WIP]: glusterd sends a SIGTERM to brick process at the time
of stopping a volume if brick_mux is not enabled.In case
of brick_mux at the time of getting a terminate signal
for last brick a brick process sends a SIGTERM to own
process for stop a brick process.The current approach
does not cleanup resources in case of either last brick
is detached or brick_mux is not enabled.

Solution: glusterd sends a terminate notification to a
brick process at the time of stopping a volume for gracefull
shutdown

Change-Id: I49b729e1205e75760f6eff9bf6803ed0dbf876ae
Fixes: #1749
Signed-off-by: Mohit Agrawal <moagrawa@redhat.com>

mohit84 committed:
core: Implement gracefull shutdown for a brick process

glusterd sends a SIGTERM to brick process at the time
of stopping a volume if brick_mux is not enabled.In case
of brick_mux at the time of getting a terminate signal
for last brick a brick process sends a SIGTERM to own
process for stop a brick process.The current approach
does not cleanup resources in case of either last brick
is detached or brick_mux is not enabled.

Solution: glusterd sends a terminate notification to a
brick process at the time of stopping a volume for gracefull
shutdown

Change-Id: I49b729e1205e75760f6eff9bf6803ed0dbf876ae
Fixes: #1749
Signed-off-by: Mohit Agrawal <moagrawa@redhat.com>

mohit84 commented:
/run regression

mohit84 commented:
/run regression

mohit84 commented:
/run regression

mohit84 commented:
/run full regression

mohit84 commented:
/run regression

mohit84 commented:
/run full regression

mohit84 commented:
/run brick-mux regression

mohit84 commented:
/run brick-mux regression

mohit84 commented:
/run brick-mux regression

mohit84 commented:
/run brick-mux regression

amarts commented:
I can confirm with this patch, I am getting operations in below sequence: (Tested for #1750)

PARENT_DOWN -> CHILD_DOWN -> fini().

mohit84 commented:
/run regression

mohit84 commented:
/run full regression

mohit84 commented:
/run brick-mux regression

amarts commented:
@mohit84 this is good I guess. Does it need something specific to pass the regression? or just a trigger again? Would be great if we can get it in.

mohit84 commented:
> @mohit84 this is good I guess. Does it need something specific to pass the regression? or just a trigger again? Would be great if we can get it in.

Yes, some more changes are required.i have to send a new patch, will upload it.

mohit84 commented:
/run regression

mohit84 commented:
/run full regression

mohit84 commented:
/run regression

mohit84 commented:
/run full regression

mohit84 commented:
/run brick-mux regression

mohit84 commented:
/run regression

mohit84 commented:
/run brick-mux regression

mohit84 commented:
/run full regression

mohit84 commented:
/run regression

mohit84 commented:
/run full regression

mohit84 commented:
/run brick-mux regression

mohit84 commented:
/run full regression

mohit84 commented:
/run regression

mohit84 commented:
/run brick-mux regression

mohit84 commented:
/run brick-mux regression

mohit84 commented:
/run regression

mohit84 commented:
/run brick-mux regression

mohit84 commented:
/run brick-mux regression

mohit84 commented:
/run regression

mohit84 commented:
/run full regression

mohit84 commented:
/run brick-mux regression

mohit84 commented:
/run full regression

mohit84 commented:
/run regression

xhernandez commented on glusterfsd/src/glusterfsd-mgmt.c:
```diff
@@ -355,41 +356,20 @@ glusterfs_handle_terminate(rpcsvc_request_t *req)
-
-        rpc_clnt_mgmt_pmap_signout(glusterfsd_ctx, xlator_req.name);
-        victim->cleanup_starting = 1;
+    if (victim->cleanup_starting)
+        goto err;
```
I think this cannot happen. if `cleanup_starting` is set, `victim` should be NULL, so we would have entered the `if` at line 346 and we'll never reach here. If we want to detect this condition, we need to do it in line 336.

mohit84 replied:
we are already doing it
if (!tmp->cleanup_starting &&
                    strcmp(tmp->name, xlator_req.name) == 0) {
                    victim = tmp;
                }

xhernandez replied:
If the volume we are trying to stop has already started cleanup, `victim->cleanup_starting` will be true, but in this case, we would have never reached line 339, so `victim` will be NULL and we'll never reach this check.

xhernandez commented on glusterfsd/src/glusterfsd-mgmt.c:
```diff
@@ -355,41 +356,20 @@ glusterfs_handle_terminate(rpcsvc_request_t *req)
-        lockflag = _gf_true;
+    rpc_clnt_mgmt_pmap_signout(glusterfsd_ctx, xlator_req.name);
+    victim->cleanup_starting = 1;
+    if (totchild == 1)
```
I'm not sure if this process is atomic enough to correctly detect the last stop of last child. For example, if there are two childs, one of them already stopping, and we want to stop the second one, `totchild` will be 2, but this is the last brick and `ctx->cleanup_starting` should be set.

How is this handled ?

mohit84 replied:
i have updated condition, totchild needs to be update only if cleanup_starting is not set.

xhernandez commented on tests/bugs/glusterd/daemon-log-level-option.t:
```diff
@@ -56,10 +57,12 @@ EXPECT 0 Trace_messages_count "/var/log/glusterfs/glustershd.log"
 # stop the volume and remove glustershd log
 TEST $CLI volume stop $V0
 rm -f /var/log/glusterfs/glustershd.log
+EXPECT_WITHIN ${PROCESS_UP_TIMEOUT} "0" online_brick_count
```
Maybe we should create a new timeout value for stopping volumes ? stop should be much faster than start, so probably we should use a smaller timeout.

xhernandez commented on xlators/protocol/server/src/server.c:
```diff
@@ -627,12 +682,40 @@ server_graph_janitor_threads(void *data)
+        {
+            totchildcount = 0;
+            for (trav_p = &top->children; *trav_p; trav_p = &(*trav_p)->next) {
+                totchildcount++;
```
We don't really need to know how many children we have. As soon as one child is detected, we can exit the loop.

mohit84 replied:
done

xhernandez reviewed (commented):
There are some minor questions regarding the changes. However I don't fully know the internals of glusterd, so I may have missed some important point (experts on glusterd should review it).

How have you tested it ? we are completely changing how things are stopped and probably we are hitting a lot of poorly tested code in the shutdown sequence of each xlator. Making sure everything is right will be difficult.

I think that it would be better to make this configurable so that we have a quick workaround in case some critical race appears. Then we can decide which is the default value based on the stability.

mohit84 commented:
/run regression

mohit84 commented:
> There are some minor questions regarding the changes. However I don't fully know the internals of glusterd, so I may have missed some important point (experts on glusterd should review it).
> 
> How have you tested it ? we are completely changing how things are stopped and probably we are hitting a lot of poorly tested code in the shutdown sequence of each xlator. Making sure everything is right will be difficult.
> 
> I think that it would be better to make this configurable so that we have a quick workaround in case some critical race appears. Then we can decide which is the default value based on the stability.

For specific to testing I am relying on regression test suite. I will test a regress testing and will update.As we know currently glusterd sends SIGTERM to a brick process in non-brick mux environment to stop the brick and in case of brick_mux a brick process sends a SIGTERM to itself if it is a last children so i believe regression test-suit itself sufficient to validate the fix.

mohit84 committed:
core: Implement gracefull shutdown for a brick process

Resolve some reviwere comment
Fixes: #1749
Signed-off-by: Mohit Agrawal <moagrawa@redhat.com>

Change-Id: I50e6a9e2ec86256b349aef5b127cc5bbf32d2561

mohit84 commented:
/query regression

mohit84 commented:
/run regression

mohit84 commented:
/run full regression

mohit84 commented:
/run brick-mux regression

mohit84 commented:
/run brick-mux regression

mohit84 commented:
/run regression

mohit84 commented:
/run regrression

mohit84 commented:
/run regression

amarts commented:
Only 3 tests failing in brick-mux:

>  3 test(s) failed 
> 22:03:20 ./tests/basic/afr/entry-self-heal-anon-dir-off.t
> 22:03:20 ./tests/basic/afr/lk-quorum.t
> 22:03:20 ./tests/bugs/distribute/bug-915554.t

See if you can get a hint by running them locally!

mohit84 commented:
> Only 3 tests failing in brick-mux:
> 
> > 3 test(s) failed
> > 22:03:20 ./tests/basic/afr/entry-self-heal-anon-dir-off.t
> > 22:03:20 ./tests/basic/afr/lk-quorum.t
> > 22:03:20 ./tests/bugs/distribute/bug-915554.t
> 
> See if you can get a hint by running them locally!

@amarts 
xavi believes we should provide a configurable option for graceful cleanup, what do you think?

amarts commented:
> xavi believes we should provide a configurable option for graceful cleanup, what do you think?

I personally believe, non-brick-mux usecase is the ideal one for gluster, and it should be properly having all shutdown etc. But considering there are so many instances with brick-mux enabled setup (with glusterfs in containers), I am OK for now to get the feature in with an option. 

We can provide this as an option for now, and cleanup further in next releases. That would be helpful.

xhernandez commented:
> > xavi believes we should provide a configurable option for graceful cleanup, what do you think?
> 
> I personally believe, non-brick-mux usecase is the ideal one for gluster, and it should be properly having all shutdown etc. But considering there are so many instances with brick-mux enabled setup (with glusterfs in containers), I am OK for now to get the feature in with an option.
> 
> We can provide this as an option for now, and cleanup further in next releases. That would be helpful.

Even in non-brick-mux use cases, I think there's a risk that some xlator doesn't correctly handle the shutdown procedure. Since we are now enforcing it, those problems could arise. It's not an issue in the change itself, which seems ok to me, but to the current implementation of PARENT_DOWN/CHILD_DOWN and fini of all xlators.

I agree that a crash during stop doesn't seem so critical as a crash during normal working, but even then it's a crash, and some environments could react to that in undesired ways, like trying to restart the process, marking the cluster as degraded, or other things that could make the problem worse.

I asked to make it configurable to have a quick workaround in case we need it. Once the functionality is considered stable and we don't detect anything on multiple different environments, we can remove that option.

I discussed this with Mohit and he will do an intensive start/stop test while I/O is ongoing to try to determine the stability of this change. If it doesn't fail after several hours, we are considering the possibility of merging this patch without that option. @amarts what do you think ?

pranithk commented:
> > > xavi believes we should provide a configurable option for graceful cleanup, what do you think?
> > 
> > 
> > I personally believe, non-brick-mux usecase is the ideal one for gluster, and it should be properly having all shutdown etc. But considering there are so many instances with brick-mux enabled setup (with glusterfs in containers), I am OK for now to get the feature in with an option.
> > We can provide this as an option for now, and cleanup further in next releases. That would be helpful.
> 
> Even in non-brick-mux use cases, I think there's a risk that some xlator doesn't correctly handle the shutdown procedure. Since we are now enforcing it, those problems could arise. It's not an issue in the change itself, which seems ok to me, but to the current implementation of PARENT_DOWN/CHILD_DOWN and fini of all xlators.
> 
> I agree that a crash during stop doesn't seem so critical as a crash during normal working, but even then it's a crash, and some environments could react to that in undesired ways, like trying to restart the process, marking the cluster as degraded, or other things that could make the problem worse.
> 
> I asked to make it configurable to have a quick workaround in case we need it. Once the functionality is considered stable and we don't detect anything on multiple different environments, we can remove that option.
> 
> I discussed this with Mohit and he will do an intensive start/stop test while I/O is ongoing to try to determine the stability of this change. If it doesn't fail after several hours, we are considering the possibility of merging this patch without that option. @amarts what do you think ?

Let's have the option. Years back we tried doing this with gfapi fini and because of bugs in the fini that are caught in different environments, we had to comment calling fini in glfsheal and make a new release.

amarts commented:
> @amarts what do you think ?

I feel, lets go with an option (which is by default 'disabled') for this patch. Once this option is there, I and others who need this option can run tests in our usecases to make sure its working OK. Also Mohit (or others doing testing) can do I/O testing after this patch, and then if found OK, we can consider making the option default to 'yes'. But that shouldn't be the reason for merging the patch itself IMO.

mohit84 commented:
> > @amarts what do you think ?
> 
> I feel, lets go with an option (which is by default 'disabled') for this patch. Once this option is there, I and others who need this option can run tests in our usecases to make sure its working OK. Also Mohit (or others doing testing) can do I/O testing after this patch, and then if found OK, we can consider making the option default to 'yes'. But that shouldn't be the reason for merging the patch itself IMO.

sure i will send a new patch with configurable option.

mohit84 committed:
core: Implement graceful shutdown for a brick process

Implement a key cluster.brick-graceful-cleanup to enable graceful
shutdown for a brick process.If key value is on glusterd sends a
detach request to stop the brick.

Fixes: #1749
Change-Id: Iba8fb27ba15cc37ecd3eb48f0ea8f981633465c3
Signed-off-by: Mohit Agrawal <moagrawa@redhat.com>

mohit84 commented:
/run regression

mohit84 commented:
/run brick-mux regression

mohit84 commented:
/run regression

mohit84 commented:
/run brick-mux regression

xhernandez commented on glusterfsd/src/glusterfsd-mgmt.c:
```diff
@@ -325,6 +330,22 @@ glusterfs_handle_terminate(rpcsvc_request_t *req)
+        }
+    }
+
+    ret = dict_get_str(dict, "cluster.brick-graceful-cleanup", &value);
```
It would be better to use a constant instead of explicit value for the name of the key.

xhernandez commented on glusterfsd/src/glusterfsd-mgmt.c:
```diff
@@ -325,6 +330,22 @@ glusterfs_handle_terminate(rpcsvc_request_t *req)
     }
     ctx = glusterfsd_ctx;
 
+    dict = dict_new();
```
`dict` check missing.

xhernandez commented on glusterfsd/src/glusterfsd-mgmt.c:
```diff
@@ -325,6 +330,22 @@ glusterfs_handle_terminate(rpcsvc_request_t *req)
+
+    ret = dict_get_str(dict, "cluster.brick-graceful-cleanup", &value);
+    if (!ret)
+        ret = gf_string2boolean(value, &enabled);
```
Shouldn't we check the returned value ?

xhernandez commented on glusterfsd/src/glusterfsd-mgmt.c:
```diff
@@ -316,7 +316,12 @@ glusterfs_handle_terminate(rpcsvc_request_t *req)
+    dict_t *dict = NULL;
+    xlator_t *this = NULL;
+    char *value = NULL;
+    gf_boolean_t enabled = _gf_false;
```
Maybe a more meaningful name would be better, like `graceful` or anything similar. `enabled` seems too generic to me and it's not clear what is enabled without code inspection.

xhernandez commented on xlators/mgmt/glusterd/src/glusterd-messages.h:
```diff
@@ -320,7 +320,8 @@ GLFS_MSGID(
     GD_MSG_ADD_BRICK_MNT_INFO_FAIL, GD_MSG_GET_MNT_ENTRY_INFO_FAIL,
     GD_MSG_QUORUM_CLUSTER_COUNT_GET_FAIL, GD_MSG_POST_COMMIT_OP_FAIL,
-    GD_MSG_POST_COMMIT_FROM_UUID_REJCT, GD_MSG_POST_COMMIT_REQ_SEND_FAIL);
+    GD_MSG_GRACEFUL_CLEANUP_SET_FAIL, GD_MSG_POST_COMMIT_FROM_UUID_REJCT,
```
New messages must be added to the end of the list. Otherwise the id's of following messages will be changed from version to version.

xhernandez commented on xlators/mgmt/glusterd/src/glusterd-op-sm.c:
```diff
@@ -2514,6 +2515,35 @@ glusterd_set_brick_mx_opts(dict_t *dict, char *key, char *value,
+    if (!strcmp(key, GLUSTERD_BRICK_GRACEFUL_CLEANUP)) {
+        ret = dict_set_dynstrn(priv->opts, GLUSTERD_BRICK_GRACEFUL_CLEANUP,
+                               SLEN(GLUSTERD_BRICK_GRACEFUL_CLEANUP),
+                               gf_strdup(value));
```
`gf_strdup()` may fail. We should check that.

xhernandez commented on xlators/mgmt/glusterd/src/glusterd-op-sm.c:
```diff
@@ -2514,6 +2515,35 @@ glusterd_set_brick_mx_opts(dict_t *dict, char *key, char *value,
+    glusterd_conf_t *priv = NULL;
+
+    this = THIS;
+    GF_VALIDATE_OR_GOTO("glusterd", this, out);
+    GF_VALIDATE_OR_GOTO(this->name, dict, out);
+    GF_VALIDATE_OR_GOTO(this->name, key, out);
+    GF_VALIDATE_OR_GOTO(this->name, value, out);
+    GF_VALIDATE_OR_GOTO(this->name, op_errstr, out);
```
Is this really necessary for an internal function ?

amarts replied:
I agree, not required.

xhernandez commented on xlators/mgmt/glusterd/src/glusterd-op-sm.c:
```diff
@@ -2514,6 +2515,35 @@ glusterd_set_brick_mx_opts(dict_t *dict, char *key, char *value,
+    xlator_t *this = NULL;
+    glusterd_conf_t *priv = NULL;
+
+    this = THIS;
```
You could pass `this` or even `this->private` or `this->private->opts` as an argument instead of using `THIS`.

amarts replied:
:+1:

xhernandez commented on xlators/mgmt/glusterd/src/glusterd-utils.c:
```diff
@@ -153,6 +153,28 @@ is_brick_mx_enabled(void)
+                        SLEN(GLUSTERD_BRICK_GRACEFUL_CLEANUP), &value);
+
+    if (!ret)
+        ret = gf_string2boolean(value, &enabled);
```
`ret` should be checked. It's ok if `dict_get_strn()` fails because it means the key is not specified and we are using the default value, but if it's present, it must be a boolean. If not, it's an error. I'm not sure if it's right to consider an error as the feature not being enabled.

amarts replied:
I feel, its better to just check for presence of key as `true`. just do `dict_get_sizen()`. Why one more conversion, and check.

mohit84 replied:
The conversion is required to take decision about enable/disable of the key. I think we should throw an error in case if it not boolean but currently we are not following it.

xhernandez commented on xlators/mgmt/glusterd/src/glusterd-utils.c:
```diff
@@ -153,6 +153,28 @@ is_brick_mx_enabled(void)
+    xlator_t *this = NULL;
+    glusterd_conf_t *priv = NULL;
+
+    this = THIS;
```
`this` can be passed as an argument, but it would even be better to pass `this->private->opts`.

xhernandez commented on xlators/mgmt/glusterd/src/glusterd-snapshot.c:
```diff
@@ -2623,7 +2623,7 @@ glusterd_do_lvm_snapshot_remove(glusterd_volinfo_t *snap_vol,
     if (gf_is_service_running(pidfile, &pid)) {
         (void)send_attach_req(this, brickinfo->rpc, brickinfo->path, NULL, NULL,
-                              GLUSTERD_BRICK_TERMINATE);
+                              GLUSTERD_BRICK_TERMINATE, 0);
```
For consistency:

```suggestion
                              GLUSTERD_BRICK_TERMINATE, _gf_false);
```

xhernandez commented on xlators/mgmt/glusterd/src/glusterd-utils.c:
```diff
@@ -6118,7 +6143,7 @@ attach_brick_callback(struct rpc_req *req, struct iovec *iov, int count,
 send_attach_req(xlator_t *this, struct rpc_clnt *rpc, char *path,
                 glusterd_brickinfo_t *brickinfo,
-                glusterd_brickinfo_t *other_brick, int op)
+                glusterd_brickinfo_t *other_brick, int op, int graceful_cleanup)
```
The argument we pass here is actually a boolean. It's better to use the same type.

```suggestion
                glusterd_brickinfo_t *other_brick, int op, gf_boolean_t graceful_cleanup)
```

xhernandez commented on xlators/mgmt/glusterd/src/glusterd-utils.c:
```diff
@@ -6154,6 +6180,33 @@ send_attach_req(xlator_t *this, struct rpc_clnt *rpc, char *path,
+            gf_smsg(this->name, GF_LOG_ERROR, errno, GD_MSG_DICT_CREATE_FAIL,
+                    NULL);
+            ret = -ENOMEM;
+            goto *errlbl;
```
Not sure how much effort this requires, but I would try to avoid indirect `goto` jumps. Makes the code much harder to follow. If it's not trivial, we can leave this for a future patch, though.

mohit84 replied:
done

xhernandez commented on xlators/mgmt/glusterd/src/glusterd-utils.c:
```diff
@@ -6264,7 +6321,7 @@ attach_brick(xlator_t *this, glusterd_brickinfo_t *brickinfo,
         if (rpc) {
             ret = send_attach_req(this, rpc, path, brickinfo, other_brick,
-                                  GLUSTERD_BRICK_ATTACH);
+                                  GLUSTERD_BRICK_ATTACH, 0);
```
```suggestion
                                  GLUSTERD_BRICK_ATTACH, _gf_false);
```

xhernandez commented on xlators/mgmt/glusterd/src/glusterd-utils.h:
```diff
@@ -214,7 +214,7 @@ glusterd_volume_stop_glusterfs(glusterd_volinfo_t *volinfo,
 send_attach_req(xlator_t *this, struct rpc_clnt *rpc, char *path,
                 glusterd_brickinfo_t *brick, glusterd_brickinfo_t *other_brick,
-                int op);
+                int op, int graceful_cleanup);
```
```suggestion
                int op, gf_boolean_t graceful_cleanup);
```

xhernandez commented on xlators/mgmt/glusterd/src/glusterd-volume-set.c:
```diff
@@ -2932,6 +2930,16 @@ struct volopt_map_entry glusterd_volopt_map[] = {
+     .value = "disable",
+     .op_version = GD_OP_VERSION_9_0,
+     .validate_fn = validate_boolean,
+     .type = GLOBAL_DOC,
```
I'm not sure if we should make this option public. I consider this experimental, so it should be enabled only by who really wants to test it. Once it's considered stable, the option will be set to "enabled" by default and shouldn't be touched by users.

amarts replied:
Ack, Ok with not being public.

xhernandez commented on xlators/mgmt/glusterd/src/glusterd-utils.c:
```diff
@@ -6213,6 +6266,10 @@ send_attach_req(xlator_t *this, struct rpc_clnt *rpc, char *path,
 err:
+    if (dict)
+        dict_unref(dict);
+    if (ret && brick_req.dict.dict_val)
```
Why it's only destroyed in case of error ? who destroys it in case of success ?

xhernandez commented on xlators/protocol/server/src/server.c:
```diff
@@ -627,12 +682,41 @@ server_graph_janitor_threads(void *data)
+        }
+        server_cleanup(this, conf);
+        glusterfs_ctx_pool_destroy(ctx);
+        exit(0);
```
If we do a complete and proper shutdown sequence, why do we need to explicitly call `exit()` here ? we could let the thread to terminate and the main program to finish normally (returning from `main()`).

mohit84 committed:
core: Implement graceful shutdown for a brick process

Resolve reviewer comments
Fixes: #1749
Signed-off-by: Mohit Agrawal <moagrawa@redhat.com>

Change-Id: I2a8eb4cf25cd8fca98d099889e4cae3954c8579e

mohit84 commented:
/run regression

mohit84 commented:
/run brick-mux regression

mohit84 commented:
/run regression

mohit84 commented:
/run smoke

mohit84 commented:
/recheck smoke

mohit84 commented:
/run brick-mux regression

mohit84 commented:
/recheck smoke

mohit84 commented:
/run regression

mohit84 commented:
/run brick-mux regression

mohit84 commented:
/run regression

mohit84 commented:
/run brick-mux regression

mohit84 commented:
/run smoke

mohit84 commented:
/run smoke

mohit84 commented:
/recheck smoke

deepshikhaaa commented:
/recheck smoke

mohit84 commented:
/recheck smoke

mohit84 commented:
/run regression

mohit84 commented:
/run brick-mux regression

mohit84 commented:
/run brick-mux regression

amarts commented on glusterfsd/src/glusterfsd-mgmt.c:
```diff
@@ -325,6 +330,28 @@ glusterfs_handle_terminate(rpcsvc_request_t *req)
+    if (!ret) {
+        ret = gf_string2boolean(value, &graceful_cleanup);
+        if (ret)
+            graceful_cleanup = _gf_false;
```
When is `graceful_cleanup` set to `_gf_true`?

mohit84 replied:
when a key (GLUSTER_BRICK_GRACEFUL_CLEANUP) is true then graceful_cleanup will be true.

mohit84 replied:
I have executed regression test-suite(https://github.com/gluster/glusterfs/pull/1892) after enabling graceful_key, it is working fine. After done regress testing we will enable the key as a default option.

mohit84 commented:
/run regression

mohit84 commented:
/run brick-mux regression

mohit84 commented:
/run regression

mohit84 commented:
/run brick-mux regression

mohit84 commented:
/run brick-mux regression

mohit84 commented:
/run brick-mux regression

mohit84 commented:
/run brick-mux regression

xhernandez commented on xlators/mgmt/glusterd/src/glusterd-op-sm.c:
```diff
@@ -2517,27 +2517,19 @@ glusterd_set_brick_mx_opts(dict_t *dict, char *key, char *value,
+            ret = -1;
+            goto out;
+        }
+        ret = dict_set_dynstrn(priv->opts, GLUSTER_BRICK_GRACEFUL_CLEANUP,
```
What happens with `dup_value` if this fails ?

mohit84 replied:
done

mohit84 committed:
core: Implement gracefull shutdown for a brick process

Resolve reviewer comment specific to avoid memory leak

Fixes: #1749
Change-Id: Ic2f09efe6190fd3776f712afc2d49b4e63de7d1f
Signed-off-by: Mohit Agrawal <moagrawa@redhat.com>

mohit84 commented:
/run regression

mohit84 commented:
/run brick-mux regression

mohit84 commented:
/run regression

mohit84 commented:
/recheck smoke

mohit84 commented:
/run regression

mohit84 commented:
/recheck smoke

mohit84 commented:
/recheck smoke

deepshikhaaa commented:
/recheck smoke

xhernandez commented on xlators/mgmt/glusterd/src/glusterd-volume-set.c:
```diff
@@ -2932,6 +2930,17 @@ struct volopt_map_entry glusterd_volopt_map[] = {
                     "brick multiplexing. Brick multiplexing ensures that "
                     "compatible brick instances can share one single "
                     "brick process."},
+    /*{.key = GLUSTER_BRICK_GRACEFUL_CLEANUP,
+     .voltype = "mgmt/glusterd",
+     .value = "disable",
+     .op_version = GD_OP_VERSION_9_0,
+     .validate_fn = validate_boolean,
+     .type = GLOBAL_DOC,
+     .description = "This global option can be used to enable/disable "
+                    "graceful cleanup, after enable graceful cleanup "
+                    "glusterd always sends a detach rpc request to stop a "
+                    "brick process."},
+    */
```
Why it's commented ? we need the option, only don't make it public with the `GLOBAL_DOC` flag...

mohit84 replied:
After doing a regress testing I will send a new patch to make it public then i will update it.

mohit84 replied:
done

mohit84 committed:
core: Implement gracefull shutdown for a brick process

Resolve reviewer comment specific to avoid memory leak

Fixes: #1749
Change-Id: I68fbbb39160a4595fb8b1b19836f44b356e89716
Signed-off-by: Mohit Agrawal <moagrawa@redhat.com>

mohit84 commented:
/run regression

mohit84 commented:
/run brick-mux regression

mohit84 commented:
/recheck smoke

mohit84 commented:
/recheck smoke

mohit84 commented:
/recheck smoke
